components/draw/SplitRouteMode.svelte

```diff
@@ -0,0 +1,143 @@
+<script>
+  import nearestPointOnLine from "@turf/nearest-point-on-line";
+  import lineSplit from "@turf/line-split";
+  import { gjScheme, map, emptyGeojson, newFeatureId } from "../../stores.js";
+  import {
+    overwriteSource,
+    overwriteLayer,
+    drawCircle,
+  } from "../../maplibre_helpers.js";
+
+  const thisMode = "split-route";
+
+  const circleRadiusPixels = 10;
+  const snapDistancePixels = 30;
+
+  export let mode;
+  export let changeMode;
+
+  export function start() {}
+  export function stop() {
+    cursor = null;
+    snappedIndex = null;
+  }
+
+  // An optional Feature<Point>
+  let cursor = null;
+  // Index into gjScheme of what we're snapped to
+  let snappedIndex = null;
+
+  $map.on("mousemove", (e) => {
+    if (mode != thisMode) {
+      return;
+    }
+
+    cursor = cursorFeature(e.lngLat.toArray(), false);
+    snappedIndex = null;
+
+    const nearbyPoint = { x: e.point.x - snapDistancePixels, y: e.point.y };
+    const thresholdKm =
+      $map.unproject(e.point).distanceTo($map.unproject(nearbyPoint)) / 1000.0;
+
+    // Are we snapped to anything?
+    let candidates = [];
+    for (let [index, f] of $gjScheme.features.entries()) {
+      if (f.geometry.type == "LineString") {
+        let snapped = nearestPointOnLine(f.geometry, cursor, {
+          units: "kilometers",
+        });
+        if (snapped.properties.dist <= thresholdKm) {
+          candidates.push([
+            index,
+            snapped.geometry.coordinates,
+            snapped.properties.dist,
+          ]);
+        }
+      }
+    }
+    candidates.sort((a, b) => a[2] - b[2]);
+
+    if (candidates.length > 0) {
+      cursor = cursorFeature(candidates[0][1], true);
+      snappedIndex = candidates[0][0];
+    }
+  });
+
+  $map.on("click", (e) => {
+    if (mode != thisMode) {
+      return;
+    }
+
+    if (snappedIndex != null) {
+      let result = lineSplit($gjScheme.features[snappedIndex], cursor);
+      if (result.features.length == 2) {
+        let piece1 = result.features[0];
+        let piece2 = result.features[1];
+
+        gjScheme.update((gj) => {
+          // Keep the old ID for one, assign a new ID to the other
+          piece1.id = gj.features[snappedIndex].id;
+          piece2.id = newFeatureId(gj);
+
+          // The properties get lost. Copy everything to both
+          piece1.properties = gj.features[snappedIndex].properties;
+          // "Deep clone"
+          piece2.properties = JSON.parse(JSON.stringify(piece1.properties));
```

Comment:
And a reminder, this round-trip is apparently the most straightforward way to "deep clone" something in JS. :\